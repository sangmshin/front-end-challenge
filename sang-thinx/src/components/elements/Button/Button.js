import React, { forwardRef } from 'react';
import { bool, func, node, object, oneOfType, string } from 'prop-types';
import cx from 'classnames';
import styles from './Button.module.scss';

const Button = forwardRef(({
  accessibleText,
  ariaLabel,
  children,
  className,
  disabled,
  onClick,
  role,
  style
}, ref) => {
  const clickHandler = e => onClick && onClick(e);

  return (
    <button
      ref={ref}
      aria-label={ariaLabel}
      className={cx('btn', className)}
      onClick={clickHandler}
      disabled={disabled}
      role={role}
      style={style}
    >
      {children}
      {accessibleText && <span className={styles.accessible}>{accessibleText}</span>}
    </button>
  )
})

Button.propTypes = {
  accessibleText: string,
  ariaLabel: string,
  children: oneOfType([
    string,
    node
  ]).isRequired,
  className: string,
  disabled: bool,
  onClick: func,
  role: string,
  style: object,
};

Button.defaultProps = {
  accessibleText: '',
  ariaLabel: '',
  children: '',
  className: '',
  disabled: false,
  onClick: _ => _,
  role: 'button',
  style: null
};

export default Button;