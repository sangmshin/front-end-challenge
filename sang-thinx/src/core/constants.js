export const COLOR_HEXCODE_MAP = {
  beige: '#f2d9a5',
  black: '#000',
}