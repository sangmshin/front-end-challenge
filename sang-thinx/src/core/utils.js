const SMALL = 768; /* Small screen / tablet */
const MEDIUM = 1024; /* Medium screen / desktop */
const LARGE = 1200; /* Large screen / wide desktop */
let BREAKPOINT;

/**
 * Finds the current matched media query breakpoint
 *
 * @returns {String} matched media query breakpoint [large, medium, small, xsmall]
 */
export function getCurrentBreakPoint() {
  if (!BREAKPOINT) {
    BREAKPOINT = {
      large: window.matchMedia(`(min-width: ${LARGE}px)`),
      medium: window.matchMedia(`(min-width: ${MEDIUM}px)`),
      small: window.matchMedia(`(min-width: ${SMALL}px)`),
      xsmall: window.matchMedia(`(max-width: ${SMALL - 1}px)`)
    };
  }

  return Object.keys(BREAKPOINT).find(breakpoint => BREAKPOINT[breakpoint].matches);
}

/**
 * Determines if the data is array and has at least one or more items
 * @return {boolean}
 */
export const isArrayWithLength = arr => Array.isArray(arr) && arr.length > 0;

/**
 * A function with no operation. (for testing)
 */
export const noop = () => null;
