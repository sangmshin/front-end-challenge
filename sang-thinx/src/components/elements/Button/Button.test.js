import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Button } from './Button';
import { noop } from 'core/utils';

const props = {
  accessibleText: '',
  ariaLabel: 'Add to cart',
  className: '',
  disabled: false,
  onClick: noop,
  role: 'button',
  style: null,
  type: 'submit'
};

describe('Component :: Button', () => {
  
  describe('#render', () => {
    const component = shallow(
      <Button {...props} >
        <div className="child" />
      </Button>
    );
    it('should match snapshot', () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });

  describe('#accessible text availability', () => {
    const component = shallow(
      <Button {...props} accessibleText='Buy Now'>
        <div className="child" />
      </Button>
    );
    it('should render "accessible text" span element in DOM', () => {
      expect(component.find('span.accessible').length).toBe(1);
    });
  });
});
