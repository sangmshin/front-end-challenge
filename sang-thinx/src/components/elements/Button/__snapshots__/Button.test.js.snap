// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component :: Button #render should match snapshot 1`] = `
<button
  aria-label="Add to cart"
  className=""
  disabled={false}
  onClick={[Function]}
  role="button"
  style={null}
  type="submit"
>
  <div
    className="child"
  />
</button>
`;
